pkg/yurtmanager/webhook/endpoints/v1/endpoints_default.go

```diff
@@ -0,0 +1,130 @@
+)
+
+// Default satisfies the defaulting webhook interface.
+func (webhook *EndpointsHandler) Default(ctx context.Context, obj runtime.Object) error {
+	endpoints, ok := obj.(*corev1.Endpoints)
+	if !ok {
+		return fmt.Errorf("expected an Endpoints object but got %T", obj)
+	}
+
+	return remapAutonomyEndpoints(ctx, webhook.Client, endpoints)
+}
+
+// isNodeAutonomous checks if the node has autonomy annotations
+// and returns true if it does, false otherwise.
+func isNodeAutonomous(ctx context.Context, c client.Client, nodeName string) (bool, error) {
+	node := &corev1.Node{}
+	err := c.Get(ctx, client.ObjectKey{Name: nodeName}, node)
+	if err != nil {
+		// If node doesn't exist, it doesn't have autonomy
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return nodeutil.IsPodBoundenToNode(node), nil
+}
+
+// isPodCrashLoopBackOff checks if the pod is crashloopbackoff
+// and returns true if it is, false otherwise.
+func isPodCrashLoopBackOff(ctx context.Context, c client.Client, podName, namespace string) (bool, error) {
+	pod := &corev1.Pod{}
+	err := c.Get(ctx, client.ObjectKey{Name: podName, Namespace: namespace}, pod)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return podutil.IsPodCrashLoopBackOff(pod.Status), nil
+}
+
+// remapAutonomyEndpoints remaps the notReadyAddresses to the readyAddresses
+// for the subsets scheduled to nodes that have autonomy annotations.
+// The function checks the pod status and if the pod is not in crashloopbackoff,
+// it remaps the address to readyAddresses.
+func remapAutonomyEndpoints(ctx context.Context, client client.Client, endpoints *corev1.Endpoints) error {
+	// Track nodes with autonomy to avoid repeated checks
+	nodesWithAutonomy := make(map[string]bool)
+
+	// Get all the notReadyAddresses for subsets
+	for i, s := range endpoints.Subsets {
+		// Create a zero-length slice with the same underlying array
+		newNotReadyAddresses := s.NotReadyAddresses[:0]
+
+		for _, a := range s.NotReadyAddresses {
+			if a.NodeName == nil || a.TargetRef == nil {
+				newNotReadyAddresses = append(newNotReadyAddresses, a)
+				continue
+			}
+
+			// Get the node and check autonomy annotations
+			hasAutonomy, ok := nodesWithAutonomy[*a.NodeName]
+			if !ok {
+				isAutonomous, err := isNodeAutonomous(ctx, client, *a.NodeName)
+				if err != nil {
+					return err
+				}
+				// Store autonomy status for future checks
+				nodesWithAutonomy[*a.NodeName] = isAutonomous
+				hasAutonomy = isAutonomous
+			}
+
+			// If the node doesn't have autonomy, skip
+			if !hasAutonomy {
+				newNotReadyAddresses = append(newNotReadyAddresses, a)
+				continue
+			}
+
+			// Get the pod
+			isPodCrashLoopBackOff, err := isPodCrashLoopBackOff(ctx, client, a.TargetRef.Name, a.TargetRef.Namespace)
+			if err != nil {
+				return err
+			}
+
+			if isPodCrashLoopBackOff {
```

Comment:
Beside CrashLoopBackOff status, are there any other status we should recognize pod as not ready?

Reply:
None that we need to handle. Terminated will fall into the normal pod lifecycle where the endpoints should be removed similar to when pod tolerationSeconds expires. Evicted also falls into this category. For an imagePull error there aren't endpoints established so this case doesn't need to be handled.

Reply:
make sense, Thanks for detailed explaination.